pkg/controllermanager/controller/managedseedset/managedseedset_actuator.go

```diff
@@ -0,0 +1,666 @@
+		}
+		setLogger.Infof("Creating shoot %s", r.getObjectName())
+		if err := r.createShoot(ctx, a.gardenClient.Client()); err != nil {
+			return status, false, err
+		}
+
+		// Increment Replicas and NextReplicaNumber in status
+		status.Replicas++
+		status.NextReplicaNumber++
+
+		return status, true, nil
+
+	case scalingIn:
+		// Determine the replica to be deleted
+		// From all deletable replicas, choose the one with lowest priority
+		if len(deletableReplicas) == 0 {
+			return status, false, fmt.Errorf("no deletable replicas found")
+		}
+		sort.Sort(ascendingPriority(deletableReplicas))
+		r := deletableReplicas[0]
+
+		// Delete the replica's managed seed (if it exists), or its shoot (if not)
+		if r.getManagedSeed() != nil {
+			setLogger.Infof("Deleting managed seed %s", r.getObjectName())
+			if err := r.deleteManagedSeed(ctx, a.gardenClient.Client()); err != nil {
+				return status, false, err
+			}
+		} else {
+			setLogger.Infof("Deleting shoot %s", r.getObjectName())
+			if err := r.deleteShoot(ctx, a.gardenClient.Client()); err != nil {
+				return status, false, err
+			}
+		}
+
+		// Decrement ReadyReplicas in status
+		status.ReadyReplicas--
+
+		return status, true, nil
+	}
+
+	return status, false, nil
+}
+
+func (a *actuator) getReplicas(ctx context.Context, set *seedmanagementv1alpha1.ManagedSeedSet) ([]*replica, error) {
+	// Convert spec.selector to labels.Selector
+	selector, err := metav1.LabelSelectorAsSelector(&set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get managed seeds, shoots, and seeds in set's namespace matching selector
+	shoots := &gardencorev1beta1.ShootList{}
+	if err := a.gardenClient.Client().List(ctx, shoots, client.InNamespace(set.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+	managedSeeds := &seedmanagementv1alpha1.ManagedSeedList{}
+	if err := a.gardenClient.Client().List(ctx, managedSeeds, client.InNamespace(set.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+	seeds := &gardencorev1beta1.SeedList{}
+	if err := a.gardenClient.Client().List(ctx, seeds, client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+
+	// Map names to objects for managed seeds and seeds
+	managedSeedsByName := make(map[string]*seedmanagementv1alpha1.ManagedSeed)
+	for i, managedSeed := range managedSeeds.Items {
+		managedSeedsByName[managedSeed.Name] = &managedSeeds.Items[i]
+	}
+	seedsByName := make(map[string]*gardencorev1beta1.Seed)
+	for i, seed := range seeds.Items {
+		seedsByName[seed.Name] = &seeds.Items[i]
+	}
+
+	// Initialize replicas
+	var replicas []*replica
+	for i, shoot := range shoots.Items {
+		// Get shoots scheduled onto this seed
+		var scheduledShoots []*gardencorev1beta1.Shoot
+		if seedsByName[shoot.Name] != nil {
+			shoots, err := a.gardenClient.GardenCore().CoreV1beta1().Shoots(metav1.NamespaceAll).List(ctx, metav1.ListOptions{
```

Comment:
I think using a meta-data only list would be better here, as you don't need the spec/status of shoots.

Reply:
Right now, we are only interested if there are scheduled shoots or not, so this could be simplified even further. However, perhaps we would like at some point to include the health of the scheduled shoots in the calculation whether to advance to the next replica, as also mentioned in the GEP. Right now this is not done because it's only interesting during updates (a newly created replica has no scheduled shoots). Having this in mind, perhaps it should stay as it is?

Reply:
As we agreed that we won't take into account scheduled shoots when determining replica readiness for now, I will do as suggested.

Reply:
@timebertt, sorry for the stupid question, but how exactly can I do a "meta-data only list"? I can't find this in APIs or generated code, and google is not helpful since I can't compose a query that doesn't return a ton of irrelevant results.

Reply:
https://github.com/gardener/gardener/blob/0f459405b52a4f94ac2960f9f46098aa33d73dd1/pkg/controllermanager/controller/project/project_control_delete.go#L68-L72

Reply:
Thanks. I tried this approach but there is an issue: reading via the `client.Client` using a field selector fails with an error `Index with name field:spec.seedName does not exist`. Reading via the `gardencoreclientset.Interface` works, but there it's not possible (or is quite unobvious) to get a `PartialObjectMetadataList` instead of a `ShootList`. Do you have any ideas?

Reply:
> reading via the client.Client using a field selector fails with an error Index with name field:spec.seedName does not exist. Reading via the

This error indicates, that the client's cache is not indexed. When using cached clients, one has to explicitly add an indexer to the cache for the specific field that shall be used in a field selector. (I currently can't find any reference usage)

But using the API reader is simpler in this case and shouldn't harm here, as you will anyways only need to read at most one shoot (`client.Limit(1)`).

Reply:
OK, great, it's done then.